### PR TITLE
Add op-deployer proof-of-concept

### PR DESCRIPTION
**Pull request by mslipper**
This PR adds a proof-of-concept for `op-deployer`, a CLI tool that allows declarative management of live OP Stack chains. This POC supports initializing the declarative chain config (called an "intent") and deploying the Superchain smart contracts using the OP Stack Manager.

An example intent for a Sepolia chain looks like this:

```toml
l1ChainID = 11155111
useFaultProofs = true
useAltDA = false
fundDevAccounts = true
contractArtifactsURL = "file:///Users/matthewslipper/dev/optimism/packages/contracts-bedrock/forge-artifacts"

[superchainRoles]
  proxyAdminOwner = "0xb9cdf788704088a4c0191d045c151fcbe2db14a4"
  protocolVersionsOwner = "0xb910764be39c84d572ff17713c615b5bfd7df650"
  guardian = "0x8c7e4a51acb17719d225bd17598b8a94b46c8767"
```

When deployed, it produces a state file that looks like this:

```json
{
  "version": 1,
  "appliedIntent": {
    "l1ChainID": 11155111,
    "superchainRoles": {
      "proxyAdminOwner": "0xb9cdf788704088a4c0191d045c151fcbe2db14a4",
      "protocolVersionsOwner": "0xb910764be39c84d572ff17713c615b5bfd7df650",
      "guardian": "0x8c7e4a51acb17719d225bd17598b8a94b46c8767"
    },
    "useFaultProofs": true,
    "useAltDA": false,
    "fundDevAccounts": true,
    "contractArtifactsURL": "file:///Users/matthewslipper/dev/optimism/packages/contracts-bedrock/forge-artifacts",
    "chains": null
  },
  "superchainDeployment": {
    "proxyAdminAddress": "0x54a6088c04a7782e69b5031579a1973a9e3c1a8c",
    "superchainConfigProxyAddress": "0xc969afc4799a9350f9f05b60748bc62f2829b03a",
    "superchainConfigImplAddress": "0x08426b74350e7cba5b52be4909c542d28b6b3962",
    "protocolVersionsProxyAddress": "0x212a023892803c7570eb317c77672c8391bf3dde",
    "protocolVersionsImplAddress": "0x2633ac74edb7ae1f1b5656e042285015f9ee477d"
  }
}
```

To use `op-deployer`, run `op-deployer init --dev --l1-chain-id <chain-id>`. This will initialize a deployment intent using the development keys in the repo. Then, run `op-deployer apply --l1-rpc-url <l1-rpc> --private-key <deployer-private-key>` to apply the deployment.

- The contracts deployment is performed by the local Go/Forge tooling.
- Upgrades of the contracts (i.e. modifying them after deploying the contracts afresh) is not currently supported. This will be supported in the future.
- The rest of the pipeline (i.e., deploying L2s and generating genesis files) is not included in this PR to keep it smaller and allow us to get buy-in on the fundamental concepts behind `op-deployer` before further implementation.

### CODE REVIEW

File: op-chain-ops/deployer/state/io.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/BurntSushi/toml"
+)
+
+func ReadJSONFile(path string, data any) error {
```
Comment (tynes) [inline op-chain-ops/deployer/state/io.go]: We have similar JSON helpers in `op-service` already, thoughts on making new ones vs importing those?

File: go.mod [hunk 2]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/consensys/gnark-crypto v0.12.1
 	github.com/crate-crypto/go-kzg-4844 v1.0.0
 	github.com/decred/dcrd/dcrec/secp256k1/v4 v4.3.0
+	github.com/docker/docker v27.2.0+incompatible
```
Comment (protolambda) [inline go.mod]: Not really a fan of pulling in the whole `docker` stack into the main go module, but it makes sense if we're doing programmatic docker work with it.

File: op-chain-ops/deployer/apply.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		l1RPCUrl := cliCtx.String(L1RPCURLFlagName)
+		workdir := cliCtx.String(WorkdirFlagName)
+		privateKey := cliCtx.String(PrivateKeyFlagName)
+
+		ctx, cancel := context.WithCancel(cliCtx.Context)
+		defer cancel()
+
+		errCh := make(chan error, 1)
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, os.Interrupt)
```
Comment (protolambda) [inline op-chain-ops/deployer/apply.go]: This is fine for now, but once op-deployer gets larger you may want to look into `op-service/cliapp` and create a service lifecycle that handles the interrupts and such.

File: op-chain-ops/deployer/docker/command.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
+package docker
+
+import (
+	"archive/tar"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/docker/docker/api/types/container"
```
Comment (protolambda) [inline op-chain-ops/deployer/docker/command.go]: `make lint-go` to catch lint issues like these. And `make lint-go-fix` to find and fix them.

File: op-chain-ops/deployer/keygen.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+type KeyGenerator interface {
+	Address(key devkeys.Key) (common.Address, error)
+	PrivateKey(key devkeys.Key) (*ecdsa.PrivateKey, error)
+}
+
+type MnemonicKeyGenerator struct {
+	dk *devkeys.MnemonicDevKeys
+}
```
Comment (protolambda) [inline op-chain-ops/deployer/keygen.go]: The `devkeys` package already provides interfaces, and the mnemonic keys approach that implement the interface. Do we really need this keygen wrapper?

File: op-chain-ops/deployer/opsm/superchain.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		return dso, fmt.Errorf("failed to create temp file for DSI: %w", err)
+	}
+	defer os.Remove(dsiInfile.Name())
+
+	if err := toml.NewEncoder(dsiInfile).Encode(opts.Input); err != nil {
+		return dso, fmt.Errorf("failed to encode DSI: %w", err)
+	}
+
+	containerDSIPath := "/opt/optimism/packages/contracts-bedrock/deployments/dsi.toml"
+	containerDSOPath := "/opt/optimism/packages/contracts-bedrock/deployments/dso.toml"
```
Comment (protolambda) [inline op-chain-ops/deployer/opsm/superchain.go]: Maybe a `/mnt/deployer/` like directory is better, to not accidentally conflict with a real contracts-bedrock file?

File: op-chain-ops/deployer/state/io.go (hunk 1 above)
Comment (protolambda) [reply op-chain-ops/deployer/state/io.go]: +1

File: op-chain-ops/deployer/opsm/superchain.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			opts.PrivateKey,
+			"--rpc-url",
+			opts.L1RPCUrl,
+			"--broadcast",
+			"--sig",
+			"run(string, string)",
+			containerDSIPath,
+			containerDSOPath,
+		),
+		docker.WithMount(dsiInfile.Name(), containerDSIPath),
```
Comment (protolambda) [inline op-chain-ops/deployer/opsm/superchain.go]: Wouldn't we want to add `docker.WithStdOut` here, to not make the deployment so opaque? If it fails somewhere, we might not understand what happened.

**Review by protolambda — APPROVED** [review]
Looks great, although I have some concerns about the platform-portability of the docker container (ending up in emulation or not), and about general docker performance. For a long-running kurtosis test it might be sufficient, but for a quick op-e2e test it would be pretty heavy. And for 3rd parties I am not sure if we want to require a full docker install, and the opaque execution of scripts (docker foundry logs and such we would want to pipe to stdout, maybe dump somewhere, etc.).

The idea of a deployer as a (input, output, apply-function) can likely be abstracted behind an interface, such that an interface takes the input, runs the thing, and then returns the output. If we want to continue using docker, as a backend for deployments. This may be great for foundry-compat, but bad for performance and toolchain size (docker, platform specifics, versioning, etc). And as discussed earlier (not entirely my idea), but for reviewers context: op-deployer ideally outputs the txs, so that the end-user can do the signing at their own convenience, rather than being hooked to an RPC for immediate tx submission, and a hotkey in docker. E.g. this way the user can simulate the tx-outcome on L1 of their chain deployment / upgrade, and use their signing process of choice.

File: op-chain-ops/deployer/state/io.go (hunk 1 above)
Comment (mslipper) [reply op-chain-ops/deployer/state/io.go]: Will import these for sure if they already exist!

File: op-chain-ops/deployer/opsm/superchain.go (hunk 7 above)
Comment (mslipper) [reply op-chain-ops/deployer/opsm/superchain.go]: It automatically capture stdout/stderr. All logs are streamed to the console.

File: op-chain-ops/deployer/opsm/superchain.go (hunk 6 above)
Comment (mslipper) [reply op-chain-ops/deployer/opsm/superchain.go]: That would require a modification to `foundry.toml` to authorize that path - happy to do so if that makes sense to you.

File: op-chain-ops/deployer/docker/command.go (hunk 4 above)
Comment (mslipper) [reply op-chain-ops/deployer/docker/command.go]: Yep, will update all linter issues prior to taking it out of draft.

**Comment by mslipper** [comment]
+1 on abstracting the container behind an interface. I'll work on that in subsequent PRs.

Candidly, I'd much rather use the Go Forge tooling - it makes things much easier, and less fragile. I'll start a thread to discuss this on Discord.

**Comment by mslipper** [comment]
Updated to use the Go/Forge tooling, and support parallel tx sending.

File: op-chain-ops/deployer/keygen.go (hunk 5 above)
Comment (mslipper) [reply op-chain-ops/deployer/keygen.go]: Removed.

File: go.mod (hunk 2 above)
Comment (mslipper) [reply go.mod]: Docker has been removed now that this supports the Go tooling.

**Pull request merged**